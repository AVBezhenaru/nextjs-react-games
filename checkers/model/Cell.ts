import { Board } from "./Board";
import { Colors } from "./Colors";
import { Figure } from "./figures/Figure";

export class Cell {
  readonly x: number;
  readonly y: number;
  readonly color: Colors;
  figure: Figure | null;
  board: Board;
  available: boolean;
  id: number;
  figureDeleteColor: boolean;

  constructor(board: Board, x: number, y: number, color: Colors, figure: Figure | null){
    this.x = x;
    this.y = y;
    this.color = color;
    this.figure = figure;
    this.board = board;
    this.figureDeleteColor = false;
    this.available = false;
    this.id = Math.random();
  }

  isEmpty(): boolean {
    return this.figure === null;
  }

  isFigure(): Figure {
    return this.figure;
  }

  //? движение королевы
  isEmptyDiagonal(target: Cell, figureElement: Figure): boolean {
    const absX = Math.abs(target.x - this.x);
    const absY = Math.abs(target.y - this.y);
    if(absY !== absX) return false;

    if(figureElement.color === 'white'){ 
      if(target.x === this.x - absX && target.y === this.y + absY ){  
        const absX2 = Math.abs(target.x - this.x);
        const absY2 = Math.abs(target.y - this.y);
        if(absY2 !== absX2) return false;

        const dy2 = this.y < target.y ? 1 : -1;
        const dx2 = this.x < target.x ? 1 : -1;
        for(let i = 1; i < absY2; i++){ 
          if(!this.board.getCell(this.x + dx2 * i, this.y + dy2 * i)?.isEmpty() && this.board.getCell(this.x + dx2 * i, this.y + dy2 * i)?.isFigure().color === 'black') {
            
            if((target.x  === (this.x + dx2 * i) - 1 && target.y === (this.y + dy2 * i) + 1 && this.board.getCell((this.x + dx2 * i) - 1, (this.y + dy2 * i) + 1)?.isEmpty() && (((this.x + dx2 * i) - 2 <=7 && (this.y + dy2 * i) + 2 <=7 ) ? (!this?.board?.getCell((this.x + dx2 * i) - 2, (this.y + dy2 * i) + 2)?.isEmpty()) : false))){
              this.board.getCell(this.x + dx2 * i, this.y + dy2 * i).isFigure().cell.figureDeleteColor = true;  
              return true;
            } else 
            if(((this.x + dx2 * i) - 2 <=7 && (this.y + dy2 * i) + 2 <=7 ) ? !this.board.getCell((this.x + dx2 * i) - 2, (this.y + dy2 * i) + 2)?.isEmpty() && this.board.getCell((this.x + dx2 * i) - 1, (this.y + dy2 * i) + 1)?.isEmpty() : false){
              return false;
            } else 
            if(!this.board.getCell((this.x + dx2 * i) - 1, (this.y + dy2 * i) + 1)?.isEmpty()){
              return false;
            }      

            this.board.getCell(this.x + dx2 * i, this.y + dy2 * i).isFigure().cell.figureDeleteColor = true; 
            return true;
          }   
          if(!this.board.getCell(this.x + dx2 * i, this.y + dy2 * i)?.isEmpty() && this.board.getCell(this.x + dx2 * i, this.y + dy2 * i)?.isFigure().color === 'white') {
            return false;
          }
        }      
        return true;
      }
      if(target.x === this.x + absX && target.y === this.y + absY ){
        const absX3 = Math.abs(target.x - this.x);
        const absY3 = Math.abs(target.y - this.y);
        if(absY3 !== absX3) return false;

        const dy3 = this.y < target.y ? 1 : -1;
        const dx3 = this.x < target.x ? 1 : -1;
        for(let i = 1; i < absY3; i++){ 

          if(!this.board.getCell(this.x + dx3  * i, this.y + dy3  * i)?.isEmpty() && this.board.getCell(this.x + dx3  * i, this.y + dy3  * i)?.isFigure().color === 'black') {
            if((target.x  === (this.x + dx3 * i) + 1 && target.y === (this.y + dy3 * i) + 1 && this.board.getCell((this.x + dx3 * i) + 1, (this.y + dy3 * i) + 1)?.isEmpty() && (((this.x + dx3 * i) + 2 <=7 && (this.y + dy3 * i) + 2 <=7 ) ? !this?.board?.getCell((this.x + dx3 * i) + 2, (this.y + dy3 * i) + 2)?.isEmpty() : false))){
              this.board.getCell(this.x + dx3 * i, this.y + dy3 * i).isFigure().cell.figureDeleteColor = true;  
              return true;
            } else 
            if(((this.x + dx3 * i) + 2 <= 7 && (this.y + dy3 * i) + 2 <= 7 ) ? !this.board.getCell((this.x + dx3 * i) + 2, (this.y + dy3 * i) + 2)?.isEmpty() && this.board.getCell((this.x + dx3 * i) + 1, (this.y + dy3 * i) + 1)?.isEmpty() : false){
              return false;
            } else 
            if(!this.board.getCell((this.x + dx3 * i) + 1, (this.y + dy3 * i) + 1)?.isEmpty()){
              return false;
            }
            this.board.getCell(this.x + dx3  * i, this.y + dy3  * i).isFigure().cell.figureDeleteColor = true; 
            return true;
          }
          if(!this.board.getCell(this.x + dx3 * i, this.y + dy3 * i)?.isEmpty() && this.board.getCell(this.x + dx3 * i, this.y + dy3 * i)?.isFigure().color === 'white') {
            return false;
          }
        }
        return true;
      }
      if(target.x === this.x - absX && target.y === this.y - absY ){
        const absX4 = Math.abs(target.x - this.x);
        const absY4 = Math.abs(target.y - this.y);
        if(absY4 !== absX4) return false;

        const dy4 = this.y < target.y ? 1 : -1;
        const dx4 = this.x < target.x ? 1 : -1;
        for(let i = 1; i < absY4; i++){ 
          if(!this.board.getCell(this.x + dx4 * i, this.y + dy4 * i)?.isEmpty() && this.board.getCell(this.x + dx4  * i, this.y + dy4  * i)?.isFigure().color === 'black') {
            
            if((target.x  === (this.x + dx4* i) - 1 && target.y === (this.y + dy4 * i) - 1 && this.board.getCell((this.x + dx4 * i) - 1, (this.y + dy4 * i) - 1)?.isEmpty() && (((this.x + dx4 * i) - 2 >= 0 && (this.y + dy4 * i) - 2 >= 0) ? (!this?.board?.getCell((this.x + dx4 * i) - 2, (this.y + dy4 * i) - 2)?.isEmpty()) : false))){
              this.board.getCell(this.x + dx4* i, this.y + dy4* i).isFigure().cell.figureDeleteColor = true;  
              return true;
            } else 
            if(((this.x + dx4 * i) - 2 >= 0 && (this.y + dy4 * i) - 2 >= 0 ) ? !this.board.getCell((this.x + dx4 * i) - 2, (this.y + dy4 * i) - 2)?.isEmpty() && this.board.getCell((this.x + dx4 * i) - 1, (this.y + dy4 * i) - 1)?.isEmpty() : false){
              console.log('h31')
              return false;
            } else 
            if(!this.board.getCell((this.x + dx4* i) - 1, (this.y + dy4* i) - 1)?.isEmpty()){
              console.log('h32')
              return false;
            }
            
            this.board.getCell(this.x + dx4 * i, this.y + dy4 * i).isFigure().cell.figureDeleteColor = true; 
            return true;
          }  
          if(!this.board.getCell(this.x + dx4 * i, this.y + dy4 * i)?.isEmpty() && this.board.getCell(this.x + dx4 * i, this.y + dy4 * i)?.isFigure().color === 'white') {
            return false;
          }
        }
        return true;
      }
      if(target.x === this.x + absX && target.y === this.y - absY ){
        const absX5 = Math.abs(target.x - this.x);
        const absY5 = Math.abs(target.y - this.y);
        if(absY5 !== absX5) return false;

        const dy5 = this.y < target.y ? 1 : -1;
        const dx5 = this.x < target.x ? 1 : -1;
        for(let i = 1; i < absY5; i++){ 
          if(!this.board.getCell(this.x + dx5 * i, this.y + dy5 * i)?.isEmpty() && this.board.getCell(this.x + dx5 * i, this.y + dy5 * i)?.isFigure().color === 'black') {
              
            if(target.x  === (this.x + dx5 * i) + 1 && target.y === (this.y + dy5 * i) - 1 && this.board.getCell((this.x + dx5 * i) + 1, (this.y + dy5 * i) - 1)?.isEmpty() && (((this.x + dx5 * i) + 2 >= 0 && (this.y + dy5 * i) - 2 >= 0) ? (!this?.board?.getCell((this.x + dx5 * i) + 2, (this.y + dy5 * i) - 2)?.isEmpty()) : false)){
              this.board.getCell(this.x + dx5 * i, this.y + dy5 * i).isFigure().cell.figureDeleteColor = true;  
              return true;
            } else 
            if(((this.x + dx5 * i) + 2 >= 0 && (this.y + dy5 * i) - 2 >= 0 ) ? !this.board.getCell((this.x + dx5 * i) + 2, (this.y + dy5 * i) - 2)?.isEmpty() && this.board.getCell((this.x + dx5 * i) + 1, (this.y + dy5 * i) - 1)?.isEmpty() : false){
              return false;
            } else 
            if(!this.board.getCell((this.x + dx5 * i) + 1, (this.y + dy5 * i) - 1)?.isEmpty()){
              return false;
            }      
            
            this.board.getCell(this.x + dx5 * i, this.y + dy5 * i).isFigure().cell.figureDeleteColor = true; 
            return true;
          }
          if(!this.board.getCell(this.x + dx5 * i, this.y + dy5 * i)?.isEmpty() && this.board.getCell(this.x + dx5 * i, this.y + dy5 * i)?.isFigure().color === 'white') {
            return false;
          }
        }
        return true;
      }
    }
    if(figureElement.color === 'black'){ 
      if(target.x === this.x - absX && target.y === this.y + absY ){  
        const absX2 = Math.abs(target.x - this.x);
        const absY2 = Math.abs(target.y - this.y);
        if(absY2 !== absX2) return false;

        const dy2 = this.y < target.y ? 1 : -1;
        const dx2 = this.x < target.x ? 1 : -1;
        for(let i = 1; i < absY2; i++){ 
          if(!this.board.getCell(this.x + dx2 * i, this.y + dy2 * i)?.isEmpty() && this.board.getCell(this.x + dx2 * i, this.y + dy2 * i)?.isFigure().color === 'white') {
            
            if((target.x  === (this.x + dx2 * i) - 1 && target.y === (this.y + dy2 * i) + 1 && this.board.getCell((this.x + dx2 * i) - 1, (this.y + dy2 * i) + 1)?.isEmpty() && (((this.x + dx2 * i) - 2 <=7 && (this.y + dy2 * i) + 2 <=7 ) ? (!this?.board?.getCell((this.x + dx2 * i) - 2, (this.y + dy2 * i) + 2)?.isEmpty()) : false))){
              this.board.getCell(this.x + dx2 * i, this.y + dy2 * i).isFigure().cell.figureDeleteColor = true;  
              return true;
            } else 
            if(((this.x + dx2 * i) - 2 <=7 && (this.y + dy2 * i) + 2 <=7 ) ? !this.board.getCell((this.x + dx2 * i) - 2, (this.y + dy2 * i) + 2)?.isEmpty() && this.board.getCell((this.x + dx2 * i) - 1, (this.y + dy2 * i) + 1)?.isEmpty() : false){
              return false;
            } else 
            if(!this.board.getCell((this.x + dx2 * i) - 1, (this.y + dy2 * i) + 1)?.isEmpty()){
              return false;
            }      

            this.board.getCell(this.x + dx2 * i, this.y + dy2 * i).isFigure().cell.figureDeleteColor = true; 
            return true;
          }   
          if(!this.board.getCell(this.x + dx2 * i, this.y + dy2 * i)?.isEmpty() && this.board.getCell(this.x + dx2 * i, this.y + dy2 * i)?.isFigure().color === 'black') {
            return false;
          }
        }      
        return true;
      }
      if(target.x === this.x + absX && target.y === this.y + absY ){
        const absX3 = Math.abs(target.x - this.x);
        const absY3 = Math.abs(target.y - this.y);
        if(absY3 !== absX3) return false;

        const dy3 = this.y < target.y ? 1 : -1;
        const dx3 = this.x < target.x ? 1 : -1;
        for(let i = 1; i < absY3; i++){ 

          if(!this.board.getCell(this.x + dx3  * i, this.y + dy3  * i)?.isEmpty() && this.board.getCell(this.x + dx3  * i, this.y + dy3  * i)?.isFigure().color === 'white') {
            if((target.x  === (this.x + dx3 * i) + 1 && target.y === (this.y + dy3 * i) + 1 && this.board.getCell((this.x + dx3 * i) + 1, (this.y + dy3 * i) + 1)?.isEmpty() && (((this.x + dx3 * i) + 2 <=7 && (this.y + dy3 * i) + 2 <=7 ) ? !this?.board?.getCell((this.x + dx3 * i) + 2, (this.y + dy3 * i) + 2)?.isEmpty() : false))){
              this.board.getCell(this.x + dx3 * i, this.y + dy3 * i).isFigure().cell.figureDeleteColor = true;  
              return true;
            } else 
            if(((this.x + dx3 * i) + 2 <= 7 && (this.y + dy3 * i) + 2 <= 7 ) ? !this.board.getCell((this.x + dx3 * i) + 2, (this.y + dy3 * i) + 2)?.isEmpty() && this.board.getCell((this.x + dx3 * i) + 1, (this.y + dy3 * i) + 1)?.isEmpty() : false){
              console.log('h21')
              return false;
            } else 
            if(!this.board.getCell((this.x + dx3 * i) + 1, (this.y + dy3 * i) + 1)?.isEmpty()){
              console.log('h22')
              return false;
            }
            this.board.getCell(this.x + dx3  * i, this.y + dy3  * i).isFigure().cell.figureDeleteColor = true; 
            return true;
          }
          if(!this.board.getCell(this.x + dx3 * i, this.y + dy3 * i)?.isEmpty() && this.board.getCell(this.x + dx3 * i, this.y + dy3 * i)?.isFigure().color === 'black') {
            return false;
          }
        }
        return true;
      }
      if(target.x === this.x - absX && target.y === this.y - absY ){
        const absX4 = Math.abs(target.x - this.x);
        const absY4 = Math.abs(target.y - this.y);
        if(absY4 !== absX4) return false;

        const dy4 = this.y < target.y ? 1 : -1;
        const dx4 = this.x < target.x ? 1 : -1;
        for(let i = 1; i < absY4; i++){ 
          if(!this.board.getCell(this.x + dx4 * i, this.y + dy4 * i)?.isEmpty() && this.board.getCell(this.x + dx4  * i, this.y + dy4  * i)?.isFigure().color === 'white') {
            
            if((target.x  === (this.x + dx4* i) - 1 && target.y === (this.y + dy4 * i) - 1 && this.board.getCell((this.x + dx4 * i) - 1, (this.y + dy4 * i) - 1)?.isEmpty() && (((this.x + dx4 * i) - 2 >= 0 && (this.y + dy4 * i) - 2 >= 0) ? (!this?.board?.getCell((this.x + dx4 * i) - 2, (this.y + dy4 * i) - 2)?.isEmpty()) : false))){
              this.board.getCell(this.x + dx4* i, this.y + dy4* i).isFigure().cell.figureDeleteColor = true;  
              return true;
            } else 
            if(((this.x + dx4 * i) - 2 >= 0 && (this.y + dy4 * i) - 2 >= 0 ) ? !this.board.getCell((this.x + dx4 * i) - 2, (this.y + dy4 * i) - 2)?.isEmpty() && this.board.getCell((this.x + dx4 * i) - 1, (this.y + dy4 * i) - 1)?.isEmpty() : false){
              console.log('h31')
              return false;
            } else 
            if(!this.board.getCell((this.x + dx4* i) - 1, (this.y + dy4* i) - 1)?.isEmpty()){
              console.log('h32')
              return false;
            }
            
            this.board.getCell(this.x + dx4 * i, this.y + dy4 * i).isFigure().cell.figureDeleteColor = true; 
            return true;
          }  
          if(!this.board.getCell(this.x + dx4 * i, this.y + dy4 * i)?.isEmpty() && this.board.getCell(this.x + dx4 * i, this.y + dy4 * i)?.isFigure().color === 'black') {
            return false;
          }
        }
        return true;
      }
      if(target.x === this.x + absX && target.y === this.y - absY ){
        const absX5 = Math.abs(target.x - this.x);
        const absY5 = Math.abs(target.y - this.y);
        if(absY5 !== absX5) return false;

        const dy5 = this.y < target.y ? 1 : -1;
        const dx5 = this.x < target.x ? 1 : -1;
        for(let i = 1; i < absY5; i++){ 
          if(!this.board.getCell(this.x + dx5 * i, this.y + dy5 * i)?.isEmpty() && this.board.getCell(this.x + dx5 * i, this.y + dy5 * i)?.isFigure().color === 'white') {
              
            if(target.x  === (this.x + dx5 * i) + 1 && target.y === (this.y + dy5 * i) - 1 && this.board.getCell((this.x + dx5 * i) + 1, (this.y + dy5 * i) - 1)?.isEmpty() && (((this.x + dx5 * i) + 2 >= 0 && (this.y + dy5 * i) - 2 >= 0) ? (!this?.board?.getCell((this.x + dx5 * i) + 2, (this.y + dy5 * i) - 2)?.isEmpty()) : false)){
              this.board.getCell(this.x + dx5 * i, this.y + dy5 * i).isFigure().cell.figureDeleteColor = true;  
              return true;
            } else 
            if(((this.x + dx5 * i) + 2 >= 0 && (this.y + dy5 * i) - 2 >= 0 ) ? !this.board.getCell((this.x + dx5 * i) + 2, (this.y + dy5 * i) - 2)?.isEmpty() && this.board.getCell((this.x + dx5 * i) + 1, (this.y + dy5 * i) - 1)?.isEmpty() : false){
              return false;
            } else 
            if(!this.board.getCell((this.x + dx5 * i) + 1, (this.y + dy5 * i) - 1)?.isEmpty()){
              return false;
            }      
            
            this.board.getCell(this.x + dx5 * i, this.y + dy5 * i).isFigure().cell.figureDeleteColor = true; 
            return true;
          }
          if(!this.board.getCell(this.x + dx5 * i, this.y + dy5 * i)?.isEmpty() && this.board.getCell(this.x + dx5 * i, this.y + dy5 * i)?.isFigure().color === 'black') {
            return false;
          }
        }
        return true;
      }
    }
    
    return false;
  }


  nullFigure() {
    this.addLostFigure(this.figure);
    if(this.figure) return this.figure = null;
  }
  
  deleteFigure(cellFigure: Cell, x: number, y: number){
    const figureDeleteColorOther = cellFigure?.figure?.color !== this.board.getCell(x,y)?.isFigure()?.color;
    if(figureDeleteColorOther){
      this.board.getCell(x,y)?.nullFigure();
    }
  }

  setFigure(figure: Figure){
    this.figure = figure;
    this.figure.cell = this;
  }

  //? добавление удаленных фигур в массив
  addLostFigure(figure: Figure){
    if(figure){
      figure?.color === Colors.BLACK ? this.board.lostBlackFigure.push(figure) : this.board.lostWhiteFigure.push(figure);
    }  
  }

  moveFigure(target: Cell){
    if(this.figure && this.figure?.canMove(target)){
      this.figure.moveFigure(target);     
      target.setFigure(this.figure);
      this.figure = null;
    } 
  }
 
}
