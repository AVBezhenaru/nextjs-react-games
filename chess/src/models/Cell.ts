/* eslint-disable @typescript-eslint/no-unused-vars */
import { Colors } from './Colors';
import { Figure, FigureNames } from './figures/Figure';
import { Board } from './Board';
import { King } from './figures/King';

export class Cell {
  readonly x: number;

  readonly y: number;

  readonly color: Colors;

  figure: Figure | null;

  board: Board;

  available: boolean; // highlight cell if there is figure and this figure is our

  underAttack: boolean;

  id: number;

  constructor(board: Board, x: number, y: number, color: Colors, figure: Figure | null) {
    this.x = x;
    this.y = y;
    this.color = color;
    this.figure = figure;
    this.board = board;
    this.available = false; // свойство у ячейки будет true если выбранная фигура может походить на эту ячейку, false - не может
    this.underAttack = false;
    this.id = Math.random();
  }

  isEmpty(): boolean {
    return this.figure === null;
  }

  isEnemy(target: Cell): boolean {
    if (target.figure) {
      return this.figure?.color !== target.figure.color;
    }
    return false;
  }

  isEmptyVertical(target: Cell): boolean {
    if (this.x !== target.x) {
      return false;
    }
    const min = Math.min(this.y, target.y);
    const max = Math.max(this.y, target.y);
    for (let y = min + 1; y < max; y++) {
      if (!this.board.getCell(this.x, y).isEmpty()) {
        return false;
      }
    }
    return true;
  }

  isEmptyHorizontal(target: Cell): boolean {
    if (this.y !== target.y) {
      return false;
    }
    const min = Math.min(this.x, target.x);
    const max = Math.max(this.x, target.x);
    for (let x = min + 1; x < max; x++) {
      if (!this.board.getCell(x, this.y).isEmpty()) {
        return false;
      }
    }
    return true;
  }

  isEmptyDiagonal(target: Cell): boolean {
    const absX = Math.abs(target.x - this.x);
    const absY = Math.abs(target.y - this.y);
    if (absY !== absX) {
      return false;
    }
    const dy = this.y < target.y ? 1 : -1;
    const dx = this.x < target.x ? 1 : -1;

    for (let i = 1; i < absY; i++) {
      if (!this.board.getCell(this.x + dx * i, this.y + dy * i).isEmpty()) {
        return false;
      }
    }
    return true;
  }

  isLastCell(figure: Figure) {
    if (figure.name === FigureNames.PAWN && (figure.cell.y === 0 || figure.cell.y === 7)) {
      this.board.transformData.shouldTransform = true;
      this.board.transformData.figure = figure;
    }
  }

  setHistoryMove(startCell: Cell, endCell: Cell, eatFigure: Figure | null) {
    const { figure } = endCell;
    const startX = startCell.x;
    const startY = startCell.y;
    const endX = endCell.x;
    const endY = endCell.y;
    const column = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h'];
    const row = ['0', '1', '2', '3', '4', '5', '6', '7'];
    const step = `${column[startX]}${Math.abs(Number(row[startY]) - 8)} - ${column[endX]}${Math.abs(
      Number(row[endY]) - 8,
    )}`;
    const check = !!(
      this.board.whiteKing?.underAttackKing || this.board.blackKing?.underAttackKing
    );
    this.board.historyMove.push({
      figureData: figure,
      moveData: step,
      eatFigureData: eatFigure?.logo,
      check,
    });
  }

  setFigure(figure: Figure) {
    this.figure = figure;
    this.figure.cell = this;
  }

  moveFigure(target: Cell) {
    let eatFigureFlag = false;
    let eatFigure = null;
    const { figure } = this; // фигура внутри ячейки
    if (figure && figure?.canMove(target)) {
      figure.moveFigure(target);
      if (target.figure) {
        this.board.addFigureToLost(target.figure);
        eatFigureFlag = true;
        eatFigure = target.figure;
      }
      target.setFigure(figure);
      this.figure = null;

      if (eatFigureFlag) this.board.checkForStalemate(); // проверка патовой ситуации

      this.isLastCell(figure);

      this.board.canAttackKing(Colors.BLACK);
      this.board.canAttackKing(Colors.WHITE);
      this.setHistoryMove(this, target, eatFigure);
      const opponentKing =
        figure.color === Colors.WHITE ? this.board.blackKing : this.board.whiteKing;
      if (opponentKing?.underAttackKing) {
        opponentKing.chekAndMateFlag = !!this.board.checkForCheckMate(opponentKing); // проверка на шах и мат
      }
    }
  }
}
