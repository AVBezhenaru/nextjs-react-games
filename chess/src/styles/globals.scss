@import url('https://fonts.googleapis.com/css2?family=Audiowide&family=Inter:wght@300;400;500;600&display=swap');

html,
body,
p {
  padding: 0;
  margin: 0;
  font-family: 'Audiowide', cursive;
  font-size: 18px;
  font-weight: 600;
  color: #646464;
}

a {
  color: inherit;
  text-decoration: none;
}

* {
  box-sizing: border-box;
}
button {
  border: none;
  background-color: transparent;
  cursor: pointer;
}

li,
ul{
  list-style: none;
  padding: 0;
  margin: 0;
}