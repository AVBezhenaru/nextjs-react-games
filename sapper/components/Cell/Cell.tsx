import React from 'react';

import classes from './Cell.module.scss';

interface ICellProps {
  classAdd: string;
  onClickLButton: () => void;
  onClickRButton: (e: React.MouseEvent<HTMLButtonElement>) => void;
  disabled: boolean;
  cellSize: number;
}

const Cell: React.FC<ICellProps> = ({
  classAdd,
  onClickLButton,
  onClickRButton,
  disabled,
  cellSize,
}) => {
  const cellStyle = {
    width: `${cellSize}px`,
    height: `${cellSize}px`,
  };

  return (
    <button
      type="button"
      className={`${classes.cell} ${classes[classAdd]}`}
      style={cellStyle}
      onClick={onClickLButton}
      onContextMenu={onClickRButton}
      disabled={disabled}
    >
      {' '}
    </button>
  );
};

export default Cell;
