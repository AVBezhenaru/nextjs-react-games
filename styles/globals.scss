@import url('https://fonts.googleapis.com/css2?family=Audiowide&family=Inter:wght@300;400;500;600&display=swap');

html,
body,
p {
  padding: 0;
  margin: 0;
  font-family: 'Audiowide', cursive;
}

a {
  color: inherit;
  text-decoration: none;
}

* {
  box-sizing: border-box;
}
