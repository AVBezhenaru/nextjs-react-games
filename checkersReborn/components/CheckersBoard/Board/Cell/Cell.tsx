import Image from 'next/image';
import { FC, memo, useMemo } from 'react';

import blackChecker from '../../../../images/blackUp.png';
import whiteChecker from '../../../../images/whiteUp.png';
import blackKingChecker from '../../../../images/blackDown.png';
import whiteKingChecker from '../../../../images/whiteDown.png';

import classes from './Cell.module.scss';
import createClasses from '../../../../helpers/ClassHelper';
import CellModel from '../../../../models/CellModel';
import { Colors } from '../../../../models/Color';
import { CheckerType } from '../../../../models/Checkers/CheckerType';
import { CheckerColor } from '../../../../models/Checkers/CheckerColor';

interface ICellProps {
  cell: CellModel,
  selected: boolean,
  click: (cell: CellModel) => void,
}

const Cell: FC<ICellProps> = ({ cell, selected, click }) => {
  const { color, checker, capturable } = cell;
  
  const cellClasses = useMemo(() => createClasses([
      [classes.Cell, true],
      [classes.Black, color === Colors.BLACK],
      [classes.White, color === Colors.WHITE],
      [classes.Captured, capturable],
      [classes.Selected, selected],
    ]
  ), [color, selected, capturable]);

  const figure = useMemo(() => {
    if (checker?.checkerType === CheckerType.CHECKER && checker?.checkerColor === CheckerColor.BLACK) return blackChecker;
    if (checker?.checkerType === CheckerType.KING && checker?.checkerColor === CheckerColor.BLACK) return blackKingChecker;
    if (checker?.checkerType === CheckerType.CHECKER && checker?.checkerColor === CheckerColor.WHITE) return whiteChecker;
    if (checker?.checkerType === CheckerType.KING && checker?.checkerColor === CheckerColor.WHITE) return whiteKingChecker;
    return null;
  }, [checker]);
  
  return (
    <div className={cellClasses} onClick={() => click(cell)}>
      {figure && <Image src={figure} className={classes.CheckerImg} />}
      {!figure && cell.available && <div className={classes.Available} />}
    </div>
  );
}

export default memo(Cell);
