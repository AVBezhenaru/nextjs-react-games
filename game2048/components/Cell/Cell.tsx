import styles from '../Cell/Cell.module.scss';
import Number from '../Number/Number';

const Cell = (props?: { number: number }) => {
  return (
    <div className={styles.cell}>{props.number ? <Number number={props.number} /> : null}</div>
  );
};

export default Cell;
