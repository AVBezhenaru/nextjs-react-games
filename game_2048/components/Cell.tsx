import React from "react";
import styles from "../../nextjs-react-games/game_2048/styles/Cell.module.css";

const Cell = () => {
  return <div className={styles.cell}></div>;
};

export default Cell;
