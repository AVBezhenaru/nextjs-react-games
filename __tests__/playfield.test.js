// import React from 'react';
// import renderer from 'react-test-renderer';
// import { expect, test, describe } from '@jest/globals';
// import { render } from '@testing-library/jest-dom';
// import { Provider } from 'react-redux';

// import PlayField from '../tetris2/components/playfield/playfield';
// import { dataSlice } from '../dataSlice';

// describe('PlayField', () => {
//   test('snapshot renders', () => {
//     const component = renderer.create(
//       <Provider dataSlice={dataSlice}>
//         <PlayField />
//       </Provider>,
//     );
//     expect(component).not.toBeNull();
//     // );
//     // const tree = component.toJSON();
//     // expect(tree).toMatchSnapshot();
//   });
// });
import { expect, it, describe } from '@jest/globals';
// import { render } from '@testing-library/jest-dom';
import renderer from 'react-test-renderer';

import Cell from '../tetris2/components/cell/cell';

describe('check cell component', () => {
  it('should look like this', () => {
    const component = renderer.create(<Cell type={1} />).toJSON();
    expect(component).toMatchSnapshot();
  });
});
