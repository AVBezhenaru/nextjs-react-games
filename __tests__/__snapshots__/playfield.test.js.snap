// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`check cell component should look like this 1`] = `
<div
  className="grid_item"
  style={
    {
      "backgroundColor": "rgba(190, 80, 120)",
    }
  }
/>
`;
